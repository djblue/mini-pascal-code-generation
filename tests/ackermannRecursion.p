program main;

class main 
begin

  function ackermann (mm, nn : integer) : integer;
  begin
    if mm < 1
    then
      ackermann := nn + 1
    else
      if nn < 1
      then
        ackermann := ackermann(mm - 1, 1)
      else
        ackermann := ackermann(mm - 1, ackermann(mm, nn - 1))
  end;

  function main;
    var res, mm, nn : integer;
  begin
    mm := 0;
    while mm < 3 do
      begin
        nn := 0;
        while nn < 5 do
          begin
            res := ackermann(mm, nn);
            print res;
            nn := nn + 1
          end;
          mm := mm + 1
      end
  end

end
.
