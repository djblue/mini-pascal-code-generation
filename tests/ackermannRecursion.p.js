var output = require('../output')
  ;

var expect = function (result, message) {
  if (!result) {
    console.log("  \033[0;31mFAIL\033[0;0m: " + message);
  } else {
    console.log("  \033[0;32mPASS\033[0;0m: " + message);
  }
};

var keys = function (obj) { return Object.keys(obj).length };

expect(keys(output) === 1, 'only one class is parsed');
expect(output.main, 'main class is parsed');

expect(keys(output.main.vars) === 0, 'no class vars for main');

expect(keys(output.main.funcs) === 2, 'only two functions');
expect(output.main.funcs.main, 'main function');
expect(output.main.funcs.ackermann, 'ackerman function');

var a = output.main.funcs.ackermann;
expect(keys(a.params) === 2, 'ackermann has two params');
expect(a.size === 0, 'offset vars correctly')
