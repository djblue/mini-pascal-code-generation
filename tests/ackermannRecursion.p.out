1
2
3
4
5
2
3
4
5
6
3
5
7
9
11
